### DISCUSSION TITLE
what the purpose of GlobalSkiaBitmapsInParallel ?

### DISCUSSION THREAD

**Discussion by Zeus64** [skia4delphi/skia4delphi#286] [Q&A] [answered]
Hello, why in Delphi Athens GlobalSkiaBitmapsInParallel is not enabled by default? What could be the reason to disable it ?

**Comment by viniciusfbb** [comment, marked as answer]
> What could be the reason to disable it.

The `TCanvasStyle.DisableGlobalLock` was introduced on the lasts days before the code freeze of the RAD 12.0, and there wouldn't be much time for beta testing. So, the `GlobalSkiaBitmapsInParallel` was introduced as experimental for now, disabled by default. I still have to review Skia's filters, but overall I think it's safe, mainly because System.Messaging is properly thread-safe now.

More information about the `GlobalSkiaBitmapsInParallel`:

> On RAD Studio 12 Athens, if your FMX app use TBitmap on background threads, you can set `GlobalSkiaBitmapsInParallel := True;` on dpr to speedup your app because it will avoid one internal lock on the UI that have on a pure FMX app. (When your thread is drawing on a TBitmap, it is locking the UI of the main thread because on the default implementation of FMX, the TCanvas can only do one drawing operation at a time.

The initial implementation of TCanvas based on Skia already considered this possibility of removing the TCanvas internal lock by adopting the use of GPU when the target is the screen (form) and the use of CPU (raster) when the target is bitmaps or printer. The most modern backends (such as Vulkan and Metal) could even support the use of the GPU in the background unlike OpenGl but this would bring much more complexity so we adopted the same as Flutter and Xamarin, the use of CPU (raster) for bitmaps (but creating a texture cache of them  when draws on screen).

**Comment by Zeus64** [comment]
Thank you, Viniciusfbb. Your thorough understanding and excellent work in this area are truly commendable. Your insights are greatly appreciated!
